config_utilities/src/external_registry.cpp

```diff
@@ -0,0 +1,217 @@
+    // are equivalent between the platforms.
+    const auto mode = boost::dll::load_mode::append_decorations | boost::dll::load_mode::search_system_folders;
+    library = boost::dll::shared_library(library_path.string(), mode);
+  }
+
+  boost::dll::shared_library library;
+};
+
+LibraryGuard::LibraryGuard() {}
+
+LibraryGuard::LibraryGuard(const std::filesystem::path& library) { libraries_.push_back(library); }
+
+LibraryGuard::~LibraryGuard() { unload(); }
+
+LibraryGuard::LibraryGuard(LibraryGuard&& other) {
+  libraries_.insert(libraries_.end(), other.libraries_.begin(), other.libraries_.end());
+  other.libraries_.clear();
+}
+
+LibraryGuard& LibraryGuard::operator=(LibraryGuard&& other) {
+  libraries_.insert(libraries_.end(), other.libraries_.begin(), other.libraries_.end());
+  other.libraries_.clear();
+  return *this;
+}
+
+void LibraryGuard::unload() {
+  for (const auto& library : libraries_) {
+    ExternalRegistry::instance().unload(library);
+  }
+
+  libraries_.clear();
+}
+
+LibraryGuard::operator bool() const { return !libraries_.empty(); }
+
+ExternalRegistry::~ExternalRegistry() {
+  std::vector<std::string> libraries;
+  for (const auto& [path, lib] : libraries_) {
+    libraries.push_back(path);
+  }
+
+  for (const auto& path : libraries) {
+    unload(path);
+  }
+}
+
+void ExternalRegistry::unload(const std::filesystem::path& library_path) {
+  if (Settings::instance().verbose_external_load) {
+    // NOTE(nathan) this is separate from the logger becuase there is no guarantee that it will be visible to the user
+    // if it is through the logger
+    std::cerr << "[WARNING] Unloading external library: " << library_path << std::endl;
+  }
+
+  auto iter = entries_.find(library_path);
+  if (iter != entries_.end()) {
+    // Remove any factories that use code from the external library in question. This will also delete any factory
+    // module maps that become empty (as they rely on types only defined in the plugin). While plugin libraries
+    // shouldn't do this, it's easy to happen in practice and will cause segfaults
+    for (const auto& entry : iter->second) {
+      internal::ModuleRegistry::removeModule(entry.key, entry.type);
+      all_entries_.erase(entry);
+    }
+  }
+
+  for (const auto& internal_ref : instances_) {
+    const auto actual_ref = internal_ref.lock();
+    if (actual_ref) {
+      actual_ref->cleanup();
+    }
+  }
+
+  entries_.erase(library_path);
+  libraries_.erase(library_path);
+  instances_.clear();
+}
+
+LibraryGuard ExternalRegistry::load(const std::filesystem::path& library_path) {
+  if (!Settings::instance().allow_external_libraries) {
+    Logger::logError("External library loading is disallowed! Not loading " + library_path.string());
+    return {};
+  }
+
+  if (Settings::instance().verbose_external_load) {
+    Logger::logInfo("Loading external library: " + library_path.string());
```

Comment:
The rest of config utilities usually wraps parsed text in apostrophes. Could be good to do here, too (and possibly for the typeInfos, too, wasn't sure whether their implementations would include them). Main reasons are:
- Consistency
- I like that it makes very clear which parts belong to the printed context if people end up using e.g. space separated strings
- There's some corner cases like trailing spaces in types etc that are really annoying to debug otherwise

```suggestion
    Logger::logInfo("Loading external library '" + library_path.string()) + "'.";
```

Reply:
Thanks, yeah I've been relying on the `ostream` operator to wrap paths (though it uses "path" instead of 'path') and just forgot when I flipped this to just use string concatenation